src/connection.cpp

```diff
@@ -119,16 +120,53 @@ void Connection::accept()
 		readTimer.expires_from_now(boost::posix_time::seconds(CONNECTION_READ_TIMEOUT));
 		readTimer.async_wait(std::bind(&Connection::handleTimeout, std::weak_ptr<Connection>(shared_from_this()), std::placeholders::_1));
 
-		// Read size of the first packet
-		boost::asio::async_read(socket,
-		                        boost::asio::buffer(msg.getBuffer(), NetworkMessage::HEADER_LENGTH),
-		                        std::bind(&Connection::parseHeader, shared_from_this(), std::placeholders::_1));
+		if (connectionState == CONNECTION_STATE_CONNECTING) {
+			// Read one-by-one ascii characters (new-line-terminated) to verify world name
+			boost::asio::async_read(socket,
+									boost::asio::buffer(msg.getBuffer(), 1),
+									std::bind(&Connection::parseProxyWorldNameIdentification, shared_from_this(), std::placeholders::_1));
+		} else {
+			// Read size of the first packet
+			boost::asio::async_read(socket,
+									boost::asio::buffer(msg.getBuffer(), NetworkMessage::HEADER_LENGTH),
+									std::bind(&Connection::parseHeader, shared_from_this(), std::placeholders::_1));
```

Comment:
Prefer lambda instead of std::bind. std::bind can be hard to read and can result in larger object files and binaries due to type information that will not be produced by equivalent lambdas.